[FEATURE] add markdown support for messages

### Description

It would be great to add markdown support so that users can easily format their messages

### Screenshots

_No response_

### Additional information

_No response_

### 👀 Have you checked if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read and understood the rules in the [Contributing Guidelines](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md)

### Are you willing to work on this issue ?

None

C:\Users\muham\Documents\GitHub\Whisper>npm install
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'server@1.0.0',
npm WARN EBADENGINE   required: { node: '16.x', npm: '8.x' },
npm WARN EBADENGINE   current: { node: 'v18.16.0', npm: '9.7.2' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'server@1.0.0',
npm WARN EBADENGINE   required: { node: '16.x', npm: '8.x' },
npm WARN EBADENGINE   current: { node: 'v18.16.0', npm: '9.7.2' }
npm WARN EBADENGINE }
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.3.2: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN cleanup Failed to remove some directories [
npm WARN cleanup   [
npm WARN cleanup     'C:\\Users\\muham\\Documents\\GitHub\\Whisper\\node_modules\\npm',
npm WARN cleanup     [Error: EPERM: operation not permitted, rmdir 'C:\Users\muham\Documents\GitHub\Whisper\node_modules\npm\node_modules\socks\build'] {
npm WARN cleanup       errno: -4048,
npm WARN cleanup       code: 'EPERM',
npm WARN cleanup       syscall: 'rmdir',
npm WARN cleanup       path: 'C:\\Users\\muham\\Documents\\GitHub\\Whisper\\node_modules\\npm\\node_modules\\socks\\build'
npm WARN cleanup     }
npm WARN cleanup   ],
npm WARN cleanup   [
npm WARN cleanup     'C:\\Users\\muham\\Documents\\GitHub\\Whisper\\node_modules\\eslint-plugin-react\\node_modules\\resolve',
npm WARN cleanup     [Error: EPERM: operation not permitted, rmdir 'C:\Users\muham\Documents\GitHub\Whisper\node_modules\eslint-plugin-react\node_modules\resolve\test\resolver\symlinked\_\symlink_target'] {
npm WARN cleanup       errno: -4048,
npm WARN cleanup       code: 'EPERM',
npm WARN cleanup       syscall: 'rmdir',
npm WARN cleanup       path: 'C:\\Users\\muham\\Documents\\GitHub\\Whisper\\node_modules\\eslint-plugin-react\\node_modules\\resolve\\test\\resolver\\symlinked\\_\\symlink_target'
npm WARN cleanup     }
npm WARN cleanup   ]
npm WARN cleanup ]
npm ERR! code 1
npm ERR! path C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-sass
npm ERR! command failed
npm ERR! command C:\Windows\system32\cmd.exe /d /s /c node scripts/build.js
npm ERR! Building: C:\Program Files\nodejs\node.exe C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\bin\node-gyp.js rebuild --verbose --libsass_ext= --libsass_cflags= --libsass_ldflags= --libsass_library=
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp verb cli [
npm ERR! gyp verb cli   'C:\\Program Files\\nodejs\\node.exe',
npm ERR! gyp verb cli   'C:\\Users\\muham\\Documents\\GitHub\\Whisper\\node_modules\\node-gyp\\bin\\node-gyp.js',
npm ERR! gyp verb cli   'rebuild',
npm ERR! gyp verb cli   '--verbose',
npm ERR! gyp verb cli   '--libsass_ext=',
npm ERR! gyp verb cli   '--libsass_cflags=',
npm ERR! gyp verb cli   '--libsass_ldflags=',
npm ERR! gyp verb cli   '--libsass_library='
npm ERR! gyp verb cli ]
npm ERR! gyp info using node-gyp@8.4.1
npm ERR! gyp info using node@18.16.0 | win32 | x64
npm ERR! gyp verb command rebuild []
npm ERR! gyp verb command clean []
npm ERR! gyp verb clean removing "build" directory
npm ERR! gyp verb command configure []
npm ERR! gyp verb find Python Python is not set from command line or npm configuration
npm ERR! gyp verb find Python Python is not set from environment variable PYTHON
npm ERR! gyp verb find Python checking if "python3" can be used
npm ERR! gyp verb find Python - executing "python3" to get executable path
npm ERR! gyp verb find Python - "python3" is not in PATH or produced an error
npm ERR! gyp verb find Python checking if "python" can be used
npm ERR! gyp verb find Python - executing "python" to get executable path
npm ERR! gyp verb find Python - executable path is "C:\Users\muham\AppData\Local\Programs\Python\Python311\python.exe"
npm ERR! gyp verb find Python - executing "C:\Users\muham\AppData\Local\Programs\Python\Python311\python.exe" to get version
npm ERR! gyp verb find Python - version is "3.11.4"
npm ERR! gyp info find Python using Python version 3.11.4 found at "C:\Users\muham\AppData\Local\Programs\Python\Python311\python.exe"
npm ERR! gyp verb get node dir no --target version specified, falling back to host node version: 18.16.0
npm ERR! gyp verb command install [ '18.16.0' ]
npm ERR! gyp verb install input version string "18.16.0"
npm ERR! gyp verb install installing version: 18.16.0
npm ERR! gyp verb install --ensure was passed, so won't reinstall if already installed
npm ERR! gyp verb install version is already installed, need to check "installVersion"
npm ERR! gyp verb got "installVersion" 9
npm ERR! gyp verb needs "installVersion" 9
npm ERR! gyp verb install version is good
npm ERR! gyp verb get node dir target node version installed: 18.16.0
npm ERR! gyp verb build dir attempting to create "build" dir: C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-sass\build
npm ERR! gyp verb build dir "build" dir needed to be created? Yes
npm ERR! gyp verb find VS msvs_version not set from command line or npm config
npm ERR! gyp verb find VS VCINSTALLDIR not set, not running in VS Command Prompt
npm ERR! gyp verb find VS could not use PowerShell to find Visual Studio 2017 or newer, try re-running with '--loglevel silly' for more details
npm ERR! gyp verb find VS looking for Visual Studio 2015
npm ERR! gyp verb find VS - not found
npm ERR! gyp verb find VS not looking for VS2013 as it is only supported up to Node.js 8
npm ERR! gyp ERR! find VS 
npm ERR! gyp ERR! find VS msvs_version not set from command line or npm config
npm ERR! gyp ERR! find VS VCINSTALLDIR not set, not running in VS Command Prompt
npm ERR! gyp ERR! find VS could not use PowerShell to find Visual Studio 2017 or newer, try re-running with '--loglevel silly' for more details
npm ERR! gyp ERR! find VS looking for Visual Studio 2015
npm ERR! gyp ERR! find VS - not found
npm ERR! gyp ERR! find VS not looking for VS2013 as it is only supported up to Node.js 8
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! find VS **************************************************************
npm ERR! gyp ERR! find VS You need to install the latest version of Visual Studio
npm ERR! gyp ERR! find VS including the "Desktop development with C++" workload.
npm ERR! gyp ERR! find VS For more information consult the documentation at:
npm ERR! gyp ERR! find VS https://github.com/nodejs/node-gyp#on-windows
npm ERR! gyp ERR! find VS **************************************************************
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! configure error
npm ERR! gyp ERR! stack Error: Could not find any Visual Studio installation to use
npm ERR! gyp ERR! stack     at VisualStudioFinder.fail (C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:122:47)
npm ERR! gyp ERR! stack     at C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:75:16
npm ERR! gyp ERR! stack     at VisualStudioFinder.findVisualStudio2013 (C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:363:14) 
npm ERR! gyp ERR! stack     at C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:71:14
npm ERR! gyp ERR! stack     at C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\lib\find-visualstudio.js:384:16
npm ERR! gyp ERR! stack     at C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\lib\util.js:54:7
npm ERR! gyp ERR! stack     at C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-gyp\lib\util.js:33:16
npm ERR! gyp ERR! stack     at ChildProcess.exithandler (node:child_process:427:5)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
npm ERR! gyp ERR! stack     at maybeClose (node:internal/child_process:1091:16)
npm ERR! gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:302:5)
npm ERR! gyp ERR! System Windows_NT 10.0.19045
npm ERR! gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\muham\\Documents\\GitHub\\Whisper\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
npm ERR! gyp ERR! cwd C:\Users\muham\Documents\GitHub\Whisper\node_modules\node-sass
npm ERR! gyp ERR! node -v v18.16.0
npm ERR! gyp ERR! node-gyp -v v8.4.1
npm ERR! gyp ERR! not ok
npm ERR! Build failed with error code: 1

npm ERR! A complete log of this run can be found in: C:\Users\muham\AppData\Local\npm-cache\_logs\2023-07-29T11_14_01_656Z-debug-0.log

C:\Users\muham\Documents\GitHub\Whisper>

--- a/client/src/components/Chat.jsx
+++ b/client/src/components/Chat.jsx
@@ -13,6 +13,7 @@ import { BiDotsVerticalRounded } from 'react-icons/bi';
 
 import { v4 as uuid } from 'uuid';
 import { debounce } from 'lodash';
+import MarkdownIt from 'markdown-it';
 
 import { useChat } from 'src/context/ChatContext';
 import { useAuth } from 'src/context/AuthContext';
@@ -71,6 +72,8 @@ const Chat = () => {
     const messageExists = (id) => {
         return Boolean(getMessage(id));
     };
+
+    const md = new MarkdownIt();
 
     useEffect(() => {
         const newMessageHandler = (message) => {
@@ -264,6 +267,8 @@ const Chat = () => {
             return;
         }
 
+        message = md.render(message);
+
         const splitMessage = message.split(' ');
         for (const word of splitMessage) {
             if (listOfBadWordsNotAllowed.includes(word)) {
@@ -402,7 +407,11 @@ const Chat = () => {
                                                 'justify-between'
                                             }`}
                                         >
-                                            <p>{message}</p>
+                                            <p
+                                                dangerouslySetInnerHTML={{
+                                                    __html: message,
+                                                }}
+                                            ></p>
                                             {sender.toString() ===
                                                 senderId.toString() &&
                                                 status !== 'pending' && (

--- a/client/src/styles/chat.css
+++ b/client/src/styles/chat.css
@@ -43,4 +43,35 @@
     .message-block .message .status {
         @apply px-[20px] text-[12px] flex gap-2 items-center;
     }
+
+    .message .content h1,
+    h2,
+    h3,
+    h4,
+    h5,
+    h6 {
+        font-size: revert !important;
+        font-weight: revert !important;
+    }
+
+    .message .content ul {
+        list-style-type: disc !important;
+        list-style-position: inside !important;
+    }
+    ol {
+        list-style: decimal !important;
+        list-style-position: inside !important;
+    }
+    ul ul,
+    ol ul {
+        list-style-type: circle !important;
+        list-style-position: inside !important;
+        margin-left: 15px !important;
+    }
+    ol ol,
+    ul ol {
+        list-style-type: lower-latin !important;
+        list-style-position: inside !important;
+        margin-left: 15px !important;
+    }
 }

